internal/controlplane/handlers_githubwebhooks.go

```diff
@@ -904,12 +978,145 @@ func (_ *Server) processInstallationAppEvent(
 		WithProviderClass(db.ProviderClassGithubApp).
 		WithPayload(payloadBytes)
 
-	res := &processingResult{
-		topic: installations.ProviderInstallationTopic,
-		iiw:   iiw,
+	return []*processingResult{
+		{
+			topic: installations.ProviderInstallationTopic,
+			iiw:   iiw,
+		},
+	}, nil
+}
+
+// processInstallationRepositoriesAppEvent processes events related to
+// changes to the list of repositories that the app can access.
+func (s *Server) processInstallationRepositoriesAppEvent(
+	ctx context.Context,
+	payload []byte,
+) ([]*processingResult, error) {
+	var event *installationRepositoriesEvent
+	if err := json.Unmarshal(payload, &event); err != nil {
+		return nil, err
+	}
+
+	// Check fields mandatory for processing the event
+	if event.GetAction() == "" {
+		return nil, errors.New("invalid event: action is nil")
+	}
+	if event.GetInstallation() == nil {
+		return nil, errors.New("invalid event: installation is nil")
+	}
+	if event.GetInstallation().GetID() == 0 {
+		return nil, errors.New("invalid installation: id is 0")
+	}
+
+	installationID := event.GetInstallation().GetID()
+	installation, err := s.store.GetInstallationIDByAppID(ctx, installationID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, fmt.Errorf("no installation found for id %d", installationID)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not determine provider id: %v", err)
+	}
+	if !installation.ProviderID.Valid {
+		return nil, errors.New("invalid provider id")
+	}
+	if !installation.ProjectID.Valid {
+		return nil, errors.New("invalid project id")
+	}
+
+	results := make([]*processingResult, 0, len(event.GetRepositoriesAdded())+len(event.GetRepositoriesRemoved()))
+	for _, repo := range event.GetRepositoriesAdded() {
+		// caveat: we're accessing the database once for every
+		// repository, which might be inefficient at scale.
+		res, err := s.repositoryAdded(
+			ctx,
+			repo,
+			event.GetAction(),
+			events.TopicQueueReconcileEntityAdd,
+			installation,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, res)
+	}
+
+	// We might want to ignore this case since we can only delete
+	// repositories there were previously registered, which would
+	// be deleted by means of "meta" and "repository" events as
+	// well.
```

Comment:
This is interesting: we run into a race condition if, for a given repository that was removed from the org, we get a `meta` before the `installation_repositories` event, or vice versa. The first deletion succeeds, while the second fails.

```
error deleting repository from DB: error retrieving repository: sql: no rows in result set
```

This is not a problem right now, as it is handled correctly by logging some errors, but we do retry the deletion a handful of times, which might cause issues at scale.

Reply:
is the `meta` event for the webhook/installation and the `installation_repositories` for the repos only? Could we process one or the other? I think in either case they could be solved by a single db lookup, no? (if we are to add and the repo is already there, then skip, if we are to delete and the repo is already deleted, skip).

Either way, nice that you caught this issue already.

Reply:
I have to test what happens if we revoke access to a repository _without_ deleting it from GitHub.

I believe that we don't get the `"meta"` event in such a case, i.e. we have to handle `"removed"` action for `"installation_repositories"` events as well.

I'll test it manually and reply here.